runtime/src/serde_snapshot.rs

```diff
@@ -367,11 +471,10 @@ where
     // but non-root stores should not be included in the snapshot
     storage.retain(|_slot, stores| !stores.is_empty());
 
-    accounts_db
-        .bank_hashes
-        .write()
-        .unwrap()
-        .insert(slot, bank_hash_info);
+    accounts_db.bank_hashes.write().unwrap().insert(
+        incremental_snapshot_slot.unwrap_or(full_snapshot_slot),
+        incremental_snapshot_bank_hash_info.unwrap_or(full_snapshot_bank_hash_info),
+    );
 
     // Process deserialized data, set necessary fields in self
     let max_id: usize = *storage
```

Comment:
Is it guaranteed that between the full snapshot generated on one node, an the incremental snapshot generated/fetched from another node, that *none* the storage entries will have overlapping ids?

Reply:
Right, yeah, I remember your comment on the other PR about the AppendVec IDs. I would assume there's no guarantee that different nodes will have the same IDs, therefore there may be overlapping IDs. I think there are a few options for this PR:

1. Re-map duplicate IDs before reconstructing the storage map.
2. Leave the implementation as-is, and then address the overlapping IDs in a later PR.
3. Prevent duplicate IDs by requiring incremental snapshots come from the same node that provides the full snapshot.

Originally I was thinking (3) would be the way to go. Maybe this is too restrictive though? How often does a node download a full snapshot? I would guess not very often, right? After a node has joined, it creates its own full snapshots and incremental snapshots for future restarts. And I think on the gossip/RPC side this is the plan; i.e. when a node asks for a snapshot and downloads it, it will be given a full snapshot and an incremental snapshot.

For (2), I think this is also my vote, regardless of (3). I tried to make this PR as small as possible, and just have the necessary bits to create and rebuild from incremental snapshots. So then (1) could be added later as necessary, when it becomes a requirement to enable getting full snapshots and incremental snapshots from different nodes.

Is that OK for this PR?

Reply:
> Maybe this is too restrictive though?

Agreed, I think this is too restrictive.

> For (2), I think this is also my vote, regardless of (3). I tried to make this PR as small as possible, and just have the necessary bits to create and rebuild from incremental snapshots. So then (1) could be added later as necessary, when it becomes a requirement to enable getting full snapshots and incremental snapshots from different nodes.

Yeah that makes sense, as long as that code makes it in before the integration with gossip it should be fine, maybe an issue to track it?

Reply:
OK, new issue has been created to track this: #18641.

Reply:
thanks!